File: python/linear_fm_waveform.py
import numpy as np
import matplotlib.pyplot as plt


class LinearFMWaveform():
    """
    Basic class to generate LFM waveform samples

    TODO: Add a way to generate multiple pulses at once 
    TODO: Add mutli-prf 
    TODO: Add an option for up and down chirping
    TODO: Add symmetric and positive sweep option (see matlab phased.LinearFMSource)
    TODO: Create a waveform namespace
    """

    def __init__(self, bandwidth=1e6, pulsewidth=100e-6, sampleRate=1e6, prf=1e3):
        self.bandwidth = bandwidth
        self.pulsewidth = pulsewidth
        self.sampleRate = sampleRate
        self.prf = prf

    def generateWaveform(self):
        """
        Generate one PRI of the LFM
        """
        wav = np.zeros((round(self.sampleRate/self.prf),), dtype=np.complex64)
        Ts = 1 / self.sampleRate
        t = np.arange(0, self.pulsewidth-Ts, Ts)
        phase = -self.bandwidth/2*t + self.bandwidth / \
            (2*self.pulsewidth)*np.square(t)
        wav[0:len(phase)] = np.exp(1j*phase)
        return wav


def main():
    """
    Plot the default waveform
    """
    lfm = LinearFMWaveform()
    wav = lfm.generateWaveform()
    plt.plot(np.real(wav))
    plt.show()


if __name__ == '__main__':
    main()

File: lib/LinearFMWaveform.cc
#include "LinearFMWaveform.h"

#include <iostream>

/*
 * Generate an LFM waveform
 */
std::vector<std::complex<float>> LinearFMWaveform::generateWaveform() {
  // TODO: Make this a constant somewhere else
  std::complex<float> j(0, 1);
  // Sample interval
  float ts = 1 / sampleRate;
  // Number of samples per pulse
  int nSamplesPulse = static_cast<int>(sampleRate * pulsewidth);
  // Number of samples per PRI
  int nSamplesPri = static_cast<int>(sampleRate / prf);
  std::vector<std::complex<float>> lfm(nSamplesPri, 0);
  float t;
  for (int n = 0; n < nSamplesPulse; n++) {
    t = n * ts;
    float phase =
        -bandwidth / 2 * t + bandwidth / (2 * pulsewidth) * std::pow(t, 2);
    lfm[n] = std::exp(j * (float)(2 * M_PI) * phase);
  }
  return lfm;
}

/*
 * Default constructor
 */
LinearFMWaveform::LinearFMWaveform() {
  bandwidth = 0;
  pulsewidth = 0;
  sampleRate = 0;
  prf = 0;
}

/*
 * Non-default constructor
 */
LinearFMWaveform::LinearFMWaveform(float bandwidth, float pulsewidth, float prf,
                                   float sampleRate) {
  this->bandwidth = bandwidth;
  this->pulsewidth = pulsewidth;
  this->prf = prf;
  this->sampleRate = sampleRate;
}

/*
 * Destructor
 */ 
LinearFMWaveform::~LinearFMWaveform() {}